koschei/backend/file_cache.py

```diff
@@ -0,0 +1,210 @@
+                    if not silent:
+                        self.log.warn(
+                            "Cannot parse cache index %s, discarding cache",
+                            e,
+                        )
+        return entries
+
+    def _write_index(self, entries):
+        """
+        Writes entries to the index file.
+        Expects it to be exclusively locked already.
+        """
+        index_path = self._p('index.json')
+        with open(index_path + '.tmp', 'w') as index:
+            json.dump(dict(version=self.INDEX_VERSION, entries=entries),
+                      index, indent=True)
+        # would need fsync (or O_PONIES) to be truly atomic, but it's just a cache
+        os.rename(index_path + '.tmp', index_path)
+
+    def _cleanup_items(self, entries, exclude):
+        """
+        Removes all directories that can be locked and don't have corresponding
+        entries in ready state.
+        Expects index to be exclusively locked. Writes to index file.
+        """
+        dirents = [d for d in os.listdir(self._cachedir)
+                   if os.path.isdir(self._p(d)) and not d == exclude]
+        for dirent in dirents:
+            if entries.get(dirent) != 'ready':
+                with FileLock(self._cachedir, dirent, exclusive=True,
+                              immediate=False) as lock:
+                    if lock.try_lock():
+                        entries.pop(dirent, None)
+                        self._write_index(entries)  # preserve atomicity
+                        self.log.info("Deleting %s", dirent)
+                        shutil.rmtree(self._p(dirent), ignore_errors=True)
+        self._write_index(entries)
+
+    def get_item(self, cache_key):
+        key = str(cache_key)
+
+        # first try getting it from the cache with shared lock only
+        # - if a process is preparing it, it will block until done
+        # - will prevent deletion
+        with FileLock(self._cachedir, key, exclusive=False):
+            with FileLock(self._cachedir, 'index', exclusive=False):
+                item = self._read_index(silent=True).get(key)
+            if item and item == 'ready':
+                # it's prepared and locked for reading
+                return self.read_item(cache_key, self._cachedir)
+
+        # we didn't get the item, or it's not complete
+        # try to get exclusive lock and start preparing
+        # it may still have been added in the meantime, so check for that as well
+        while True:
+            with FileLock(self._cachedir, key, exclusive=True):
+                with FileLock(self._cachedir, 'index', exclusive=True) as index_lock:
+                    entries = self._read_index()
+                    item = entries.get(key)
+                    if item and item == 'ready':
+                        # other process added it in the meantime
+                        index_lock.unlock()
+                        return self.read_item(cache_key, self._cachedir)
```

Comment:
I haven't found any documentation specifying that relocking is atomic, do you know if it is? If it's not atomic, it would require additional boilerplate.